### PR TITLE
Added simple framework for git hooks, + checkstyle hooks

### PR DESCRIPTION
**Pull request by zhongfu**
As the title suggests, there's only one pre-commit hook turned on now, but we can add more pre-commit hooks easily or extend this to other available hooks (e.g. pre-push)

To turn them on *nix:

```
rm -r .git/hooks/
ln -s ../.githooks .git/hooks
```

or on Windows:

```
rmdir /S /Q .git/hooks/
mklink /D .githooks .git/hooks
```

### CODE REVIEW

File: .githooks/pre-commit.d/checkstyle [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+case "${unameOut}" in
+    Linux*)     machine=Linux;;
+    Darwin*)    machine=Mac;;
+    CYGWIN*)    machine=Cygwin;;
+    MINGW*)     machine=MinGw;;
+    *)          machine="UNKNOWN:${unameOut}"
+esac
+
+if [[ "$machine" == "Cygwin" || "$machine" == "MinGw" ]]; then
+    ./gradlew.bat checkstyleMain
```
Comment (ian-from-dover) [inline .githooks/pre-commit.d/checkstyle:15]: Perhaps we could run `checkstyleTest` as well by doing `./gradlew.bat check coverage`. same for line 18 below

**Review by ian-from-dover — COMMENTED** [review]
1 minor change suggested

File: .githooks/pre-commit.d/checkstyle (hunk 1 above)
Comment (Spyobird) [reply .githooks/pre-commit.d/checkstyle:15]: I think only running `check` is needed

**Comment by zhongfu** [comment]
also added a lil touch of formatting

image

**Review by Spyobird — APPROVED** [review]
lgtm

**Comment by Spyobird** [comment]
> As the title suggests, there's only one pre-commit hook turned on now, but we can add more pre-commit hooks easily or extend this to other available hooks (e.g. pre-push)
> 
> To turn them on *nix:
> 
> ```
> rm -r .git/hooks/
> ln -s ../.githooks .git/hooks
> ```
> 
> or on Windows:
> 
> ```
> rmdir /S /Q .git/hooks/
> mklink /D .githooks .git/hooks
> ```

Correction for Windows:
```
rmdir /S /Q .git\hooks\
mklink /D .git\hooks .githooks 
```

**Pull request merged**